torch/optim/_adafactor.py

```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
+
+    grouped_tensors = _group_tensors_by_device_dtype_and_is_multidim(
+        [params, grads, row_vars, col_vars, variances, state_steps]  # type: ignore[list-item]
+    )
+    for (_, dtype, is_multidim), (
+        (
+            device_params,
+            device_grads,
+            device_row_vars,
+            device_col_vars,
+            device_variances,
+            device_state_steps,
+        )
+    ) in grouped_tensors.items():
+        if eps1 is None:
+            eps1 = torch.finfo(dtype).eps
+
+        if TYPE_CHECKING:
+            assert device_state_steps[0] is not None
+
+        if maximize:
+            device_grads = torch._foreach_neg(device_grads)  # type: ignore[assignment,arg-type]
+
+        # Update steps
+        # If steps are on CPU, foreach will fall back to the slow path, which is a for-loop calling t.add(1) over
+        # and over. 1 will then be wrapped into a Tensor over and over again, which is slower than if we just
+        # wrapped it once now. The alpha is required to assure we go to the right overload.
+        if not torch._utils.is_compiling() and device_state_steps[0].is_cpu:
+            torch._foreach_add_(
+                device_state_steps, torch.tensor(1.0, device="cpu"), alpha=1.0  # type: ignore[arg-type]
+            )
+        else:
+            torch._foreach_add_(device_state_steps, 1)  # type: ignore[arg-type]
+
+        one_minus_beta2_ts = []
+        beta2_ts = []
+        rho_ts = []
+        for s in device_state_steps:
+            one_minus_beta2_ts.append(s.item() ** beta2_decay)  # type: ignore[union-attr]
+            beta2_ts.append(1 - s.item() ** beta2_decay)  # type: ignore[union-attr]
+            rho_ts.append(min(lr, 1 / (s.item() ** 0.5)))  # type: ignore[union-attr]
+
+        alphas = [
+            max(eps2, p.norm(2).item() / (p.numel() ** 0.5)) * r  # type: ignore[union-attr]
+            for p, r in zip(device_params, rho_ts)
+        ]
+
+        # Perform stepweight decay
+        if weight_decay != 0:
+            torch._foreach_mul_(device_params, 1 - lr * weight_decay)  # type: ignore[arg-type]
+
+        if is_multidim:
+            assert (
+                device_row_vars[0] is not None and device_col_vars[0] is not None
+            ), "row_var and col_var should be defined when grad is multidimensional"
+            # same as (g * g).mean(dim=-1) w/o materializing an intermediate size g
+            row_means = [
+                torch.norm(grad, dim=-1, keepdim=True) for grad in device_grads
+            ]
+            torch._foreach_mul_(row_means, row_means)
+            torch._foreach_div_(row_means, [grad.size(-1) for grad in device_grads])  # type: ignore[union-attr]
+            torch._foreach_mul_(device_row_vars, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(row_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_row_vars, row_means)  # type: ignore[arg-type]
```

Comment:
in the future would be a 
```
torch._foreach_lerp_(device_row_vars, row_means, one_minus_beta2_ts)
```

if we had ScalarList support for _foreach_lerp 3rd arg

Reply:
Do you expect to do these before merging the PR?
Should these improvements be recapped in an issue?